dotnet/sdk-phase-1.md

```diff
@@ -0,0 +1,1052 @@
+
+Doing the repo rename was deemed easier than going through manually closing issues, closing PRs, creating empty `main`
+branch, and keeping `master` branch around for those reading older code.
+
+## Namespace/Package
+
+* Will publish as `Temporalio`
+  * Need to reserve prefix `Temporalio` on NuGet, see
+    https://learn.microsoft.com/en-us/nuget/nuget-org/id-prefix-reservation
+* Namespace will be `Temporal`
+  * See discussion topics later about this being `Temporal.Sdk` instead
+
+## Workflow Definition
+
+Here's a definition for an other-language-defined workflow (from Python README):
+
+```csharp
+namespace MyNamespace;
+
+using System.Text.Json.Serialization;
+using Temporal.Workflow;
+
+public record GreetingInfo
+{
+    [JsonPropertyName("salutation")]
+    public string Salutation { get; init; } = "Hello";
+
+    [JsonPropertyName("name")]
+    public string Name { get; init; } = "<unknown>";
+}
+
+[Workflow]
+public interface IGreetingWorkflow
+{
+    static readonly IGreetingWorkflow Ref = Refs<IGreetingWorkflow>.Instance;
+
+    [WorkflowRun]
+    Task<string> RunAsync(GreetingInfo info);
+
+    [WorkflowSignal]
+    Task UpdateInfoAsync(GreetingInfo info);
+
+    [WorkflowSignal]
+    Task CompleteWithGreetingAsync();
+
+    [WorkflowQuery]
+    string CurrentGreeting();
+}
+```
+
+Notes:
+
+* We use regular .NET JSON serialization (see data conversion later)
+* Workflow attributes, and in the future workflow runtime features, are in the `Temporal.Workflow` namespace
+* We only support workflow interfaces right now because we don't have a workflow implementation yet
+  * For now, everything non-static must have some form of `[Workflow` attribute, and not have a default implementation
+* Every workflow interface must have the non-inheritable `Workflow` attribute
+  * Optional `Name` positional property can be given. Default is the unqualified type name with the "I" prefix removed
+    if it's an interface (which it has to be currently) and the next letter is also capital
+    * So what does `IDWorkflow` become? Yes, `DWorkflow` :-( Weighed as an acceptably rare cost for obvious benefit
+* Every workflow must explicitly have one and only one method with a non-inheritable `WorkflowRun` attribute
+  * Can be in base class, but must also be overridden in this workflow class w/ the attribute even if it just delegates
+    to super
+  * Method must return a `Task`
+* Workflow signals are methods with the non-inheritable `WorkflowSignal` attribute
+  * Can be in base class, but if overridden, must also have the attribute
+  * Method must return a `Task` and can accept arguments
+  * Optional `Name` positional property can be given. Default is the unqualified method name with "Async" trimmed off if
+    present.
+* Workflow queries are methods with the non-inheritable `WorkflowQuery` attribute
+  * Can be in base class, but if overridden, must also have the attribute
+  * Method cannot return a `Task`, but can accept arguments
```

Comment:
Yes, I believe so. I have to confirm as I did with activities that method references are properly disambiguated to use the more specific overload when inferred, but otherwise I think we can.

Reply:
>technically there's no reason why query handlers can not call external non-workflow APIs to respond using both workflow state and external data

Do we support async query handlers in any lang already? I'm probably missing some context or knowledge here, but it seems to me like there are quite a few things that makes this undesirable (ie. doesn't a query has to complete in the same WFT?), and even impossible in sandboxed langs...

Reply:
> Do we support async query handlers in any lang already?

Yes, Python. When originally designed, the concern was that there were async things in the language that a query handler might want to do. In retrospect though after using for a while, I didn't really find any `async` constructs that one would use in a query handler. So I may want to remove it, we'll see.

> seems to me like there are quite a few things that makes this undesirable (ie. doesn't a query has to complete in the same WFT?), and even impossible in sandboxed langs

Just because a query has to complete in the same WFT doesn't mean it can't use something async in the language. Some utilities in languages that are effectively immediate are marked async and in some languages you can't even invoke them outside of an async context. Having said that, in most languages invoking one of these even outside of an async context is often doable.